File: src/oraculus_di_auditor/analysis/cross_reference.py
"""Cross-Reference Auditor for Legal Documents.

Detects cross-jurisdiction references and potential conflicts between
different legal corpora (federal vs state, USC vs CFR, etc.).

Author: GitHub Copilot Agent
Date: 2025-11-13
"""

import re
from typing import Any

# Pattern definitions for common legal citations
CITATION_PATTERNS = {
    "usc": r"\b\d+\s+U\.?S\.?C\.?\s+§?\s*\d+",  # e.g., "42 U.S.C. § 1983"
    "cfr": r"\b\d+\s+C\.?F\.?R\.?\s+§?\s*\d+",  # e.g., "21 CFR § 50.25"
    "cal_code": r"\bCal\.?\s+\w+\.?\s+Code\b",  # e.g., "Cal. Penal Code"
    "pub_law": r"\bPub\.?\s*L\.?\s+No\.?\s+\d+-\d+",  # e.g., "Pub. L. No. 111-148"
    "stat": r"\b\d+\s+Stat\.?\s+\d+",  # e.g., "124 Stat. 119"
}


def detect_cross_jurisdiction_refs(text: str) -> list[dict[str, Any]]:
    """Detect references across different legal jurisdictions.

    Args:
        text: Document text to analyze

    Returns:
        List of detected cross-jurisdiction references with details
    """
    references = []

    # Check for each citation type
    detected_types = {}
    for cite_type, pattern in CITATION_PATTERNS.items():
        matches = re.finditer(pattern, text, re.IGNORECASE)
        detected = [match.group() for match in matches]
        if detected:
            detected_types[cite_type] = detected

    # Identify cross-jurisdiction patterns
    if "usc" in detected_types and "cal_code" in detected_types:
        references.append(
            {
                "type": "federal_state_cross_reference",
                "federal": detected_types["usc"],
                "state": detected_types["cal_code"],
                "severity": "info",
                "description": (
                    "Document contains both federal (USC) and "
                    "California state code references"
                ),
            }
        )

    if "cfr" in detected_types and "cal_code" in detected_types:
        references.append(
            {
                "type": "cfr_state_cross_reference",
                "federal": detected_types["cfr"],
                "state": detected_types["cal_code"],
                "severity": "info",
                "description": (
                    "Document contains both federal regulations (CFR) and "
                    "California state code references"
                ),
            }
        )

    return references


def cross_reference_audit(docs: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Audit documents for cross-jurisdiction references and anomalies.

    Args:
        docs: List of document dictionaries with 'id', 'text',
            and optional 'jurisdiction'

    Returns:
        List of anomaly findings with document ID and issue details
    """
    anomalies = []

    for doc in docs:
        doc_id = doc.get("id", "unknown")
        text = doc.get("text", "")
        jurisdiction = doc.get("jurisdiction", "unknown")

        # Detect cross-jurisdiction references
        cross_refs = detect_cross_jurisdiction_refs(text)

        for ref in cross_refs:
            anomalies.append(
                {
                    "id": doc_id,
                    "jurisdiction": jurisdiction,
                    "issue": ref["type"],
                    "severity": ref["severity"],
                    "description": ref["description"],
                    "details": {
                        k: v
                        for k, v in ref.items()
                        if k not in ["type", "severity", "description"]
                    },
                }
            )

        # Additional checks for specific jurisdiction mismatches
        if jurisdiction == "federal":
            # Federal document shouldn't primarily reference state codes
            state_matches = re.findall(
                CITATION_PATTERNS["cal_code"], text, re.IGNORECASE
            )
            federal_matches = len(
                re.findall(CITATION_PATTERNS["usc"], text, re.IGNORECASE)
            ) + len(re.findall(CITATION_PATTERNS["cfr"], text, re.IGNORECASE))

            if state_matches and len(state_matches) > federal_matches:
                anomalies.append(
                    {
                        "id": doc_id,
                        "jurisdiction": jurisdiction,
                        "issue": "jurisdiction_mismatch",
                        "severity": "warning",
                        "description": (
                            f"Federal document contains more state references "
                            f"({len(state_matches)}) than federal references "
                            f"({federal_matches})"
                        ),
                    }
                )

        elif jurisdiction in ["california", "state"]:
            # State document with primarily federal references might be misclassified
            federal_matches = len(
                re.findall(CITATION_PATTERNS["usc"], text, re.IGNORECASE)
            ) + len(re.findall(CITATION_PATTERNS["cfr"], text, re.IGNORECASE))
            state_matches = re.findall(
                CITATION_PATTERNS["cal_code"], text, re.IGNORECASE
            )

            if federal_matches > len(state_matches) * 2:
                anomalies.append(
                    {
                        "id": doc_id,
                        "jurisdiction": jurisdiction,
                        "issue": "jurisdiction_mismatch",
                        "severity": "warning",
                        "description": (
                            f"State document contains significantly more "
                            f"federal references ({federal_matches}) than "
                            f"state references ({len(state_matches)})"
                        ),
                    }
                )

    return anomalies

File: mobile/lib/analysis/detectors/crossReference.ts
/**
 * Cross-Reference Auditor for Legal Documents.
 *
 * Detects cross-jurisdiction references and potential conflicts between
 * different legal corpora.
 * Direct port of src/oraculus_di_auditor/analysis/cross_reference.py
 */

import { CrossReference, CrossReferenceAnomaly, NormalizedDocument } from '../types';

/** Pattern definitions for common legal citations */
export const CITATION_PATTERNS: Record<string, RegExp> = {
  usc: /\b\d+\s+U\.?S\.?C\.?\s+§?\s*\d+/gi,
  cfr: /\b\d+\s+C\.?F\.?R\.?\s+§?\s*\d+/gi,
  cal_code: /\bCal\.?\s+\w+\.?\s+Code\b/gi,
  pub_law: /\bPub\.?\s*L\.?\s+No\.?\s+\d+-\d+/gi,
  stat: /\b\d+\s+Stat\.?\s+\d+/gi,
};

/**
 * Detect references across different legal jurisdictions.
 * Mirrors Python: detect_cross_jurisdiction_refs(text)
 */
export function detectCrossJurisdictionRefs(text: string): CrossReference[] {
  const references: CrossReference[] = [];

  // Check for each citation type
  const detectedTypes: Record<string, string[]> = {};
  for (const [citeType, pattern] of Object.entries(CITATION_PATTERNS)) {
    const regex = new RegExp(pattern.source, 'gi');
    const matches: string[] = [];
    let match: RegExpExecArray | null;
    while ((match = regex.exec(text)) !== null) {
      matches.push(match[0]);
    }
    if (matches.length > 0) {
      detectedTypes[citeType] = matches;
    }
  }

  // Identify cross-jurisdiction patterns
  if ('usc' in detectedTypes && 'cal_code' in detectedTypes) {
    references.push({
      type: 'federal_state_cross_reference',
      federal: detectedTypes.usc,
      state: detectedTypes.cal_code,
      severity: 'info',
      description:
        'Document contains both federal (USC) and California state code references',
    });
  }

  if ('cfr' in detectedTypes && 'cal_code' in detectedTypes) {
    references.push({
      type: 'cfr_state_cross_reference',
      federal: detectedTypes.cfr,
      state: detectedTypes.cal_code,
      severity: 'info',
      description:
        'Document contains both federal regulations (CFR) and California state code references',
    });
  }

  return references;
}

/**
 * Audit documents for cross-jurisdiction references and anomalies.
 * Mirrors Python: cross_reference_audit(docs)
 */
export function crossReferenceAudit(
  docs: NormalizedDocument[]
): CrossReferenceAnomaly[] {
  const anomalies: CrossReferenceAnomaly[] = [];

  for (const doc of docs) {
    const docId = doc.id || 'unknown';
    const text = doc.text || '';
    const jurisdiction = doc.jurisdiction || 'unknown';

    // Detect cross-jurisdiction references
    const crossRefs = detectCrossJurisdictionRefs(text);

    for (const ref of crossRefs) {
      const details: Record<string, unknown> = {};
      for (const [k, v] of Object.entries(ref)) {
        if (k !== 'type' && k !== 'severity' && k !== 'description') {
          details[k] = v;
        }
      }

      anomalies.push({
        id: docId,
        jurisdiction,
        issue: ref.type,
        severity: ref.severity,
        description: ref.description,
        details,
      });
    }

    // Additional checks for specific jurisdiction mismatches
    if (jurisdiction === 'federal') {
      const calCodePattern = new RegExp(CITATION_PATTERNS.cal_code.source, 'gi');
      const uscPattern = new RegExp(CITATION_PATTERNS.usc.source, 'gi');
      const cfrPattern = new RegExp(CITATION_PATTERNS.cfr.source, 'gi');

      const stateMatches = text.match(calCodePattern) || [];
      const federalMatches =
        (text.match(uscPattern) || []).length +
        (text.match(cfrPattern) || []).length;

      if (stateMatches.length > 0 && stateMatches.length > federalMatches) {
        anomalies.push({
          id: docId,
          jurisdiction,
          issue: 'jurisdiction_mismatch',
          severity: 'warning',
          description: `Federal document contains more state references (${stateMatches.length}) than federal references (${federalMatches})`,
        });
      }
    } else if (jurisdiction === 'california' || jurisdiction === 'state') {
      const uscPattern = new RegExp(CITATION_PATTERNS.usc.source, 'gi');
      const cfrPattern = new RegExp(CITATION_PATTERNS.cfr.source, 'gi');
      const calCodePattern = new RegExp(CITATION_PATTERNS.cal_code.source, 'gi');

      const federalMatches =
        (text.match(uscPattern) || []).length +
        (text.match(cfrPattern) || []).length;
      const stateMatches = text.match(calCodePattern) || [];

      if (federalMatches > stateMatches.length * 2) {
        anomalies.push({
          id: docId,
          jurisdiction,
          issue: 'jurisdiction_mismatch',
          severity: 'warning',
          description: `State document contains significantly more federal references (${federalMatches}) than state references (${stateMatches.length})`,
        });
      }
    }
  }

  return anomalies;
}
